scylla/src/transport/locator/tablets.rs

```diff
@@ -0,0 +1,630 @@
+                Some((r, *shard as Shard))
+            } else {
+                // TODO: Should this be an error? When can this happen?
+                warn!("Node {replica} from system.tablets not present in ClusterData.known_peers. Skipping this replica");
+                None
+            })
+            .collect();
+
+        let mut per_dc: HashMap<String, Vec<(Arc<Node>, Shard)>> = HashMap::new();
+        all.iter().for_each(|(replica, node)| {
+            if let Some(dc) = replica.datacenter.as_ref() {
+                if let Some(replicas) = per_dc.get_mut(dc) {
+                    replicas.push((Arc::clone(replica), *node));
+                } else {
+                    per_dc.insert(dc.to_string(), vec![(Arc::clone(replica), *node)]);
+                }
+            }
+        });
+
+        Self { all, per_dc }
+    }
+}
+
+impl TableTablets {
+    fn tablet_for_token(&self, token: Token) -> Option<&Tablet> {
+        let idx = self
+            .tablet_list
+            .partition_point(|tablet| tablet.last_token < token);
+        let tablet = self.tablet_list.get(idx);
+        tablet.filter(|t| t.first_token <= token)
+    }
+
+    pub(crate) fn replicas_for_token(&self, token: Token) -> Option<&[(Arc<Node>, Shard)]> {
+        self.tablet_for_token(token)
+            .map(|tablet| tablet.replicas.all.as_ref())
+    }
+
+    pub(crate) fn dc_replicas_for_token(
+        &self,
+        token: Token,
+        dc: &str,
+    ) -> Option<&[(Arc<Node>, Shard)]> {
+        self.tablet_for_token(token).map(|tablet| {
+            tablet
+                .replicas
+                .per_dc
+                .get(dc)
+                .map(|x| x.as_slice())
+                .unwrap_or(&[])
+        })
+    }
+
+    fn add_tablet(&mut self, tablet: Tablet) {
+        // Smallest `left_idx` for which `tablet.first_token` is LESS OR EQUAL to `tablet_list[left_idx].last_token`.
+        // It implies that `tablet_list[left_idx]` overlaps with `tablet` iff `tablet.last_token`
+        // is GREATER OR EQUAL to `tablet_list[left_idx].first_token`.
+        let left_idx = self
+            .tablet_list
+            .partition_point(|t| t.last_token < tablet.first_token);
+        // Smallest `right_idx` for which `tablet.last_token` is LESS than `tablet_list[right_idx].first_token`.
+        // It means that `right_idx` is the index of first tablet that is "to the right" of `tablet` and doesn't overlap with it.
+        // From this it follows that if `tablet_list[left_idx]` turns out to not overlap with `tablet`, then `left_idx == right_idx`
+        // and we won't remove any tablets because `tablet` doesn't overlap with any existing tablets.
+        let right_idx = self
+            .tablet_list
+            .partition_point(|t| t.first_token <= tablet.last_token);
+        self.tablet_list.drain(left_idx..right_idx);
+        self.tablet_list.insert(left_idx, tablet);
+    }
+}
```

Comment:
I don't exactly remember - I think @avelanarius said something about a maximum of 100k per table.
We could use some tree-like structure instead of Vec for faster insertion / deletion, possibly at the cost of slower lookup / increased memory usage.
Maybe it's possible to write this with `BTreeMap` (if not we'd need to use external crate / hand written data structure), but first we'd need some benchmarks to see if it is really faster.

Fortunately the implementation doesn't leak outside this module so we can easily change this later.

Reply:
This PR is already quite big, so I think it's best to postpone evaluating different implementation. When we merge this I'll open an issue about benchmarking tree-based implementation.

Reply:
Opened: https://github.com/scylladb/scylla-rust-driver/issues/983